File: rust/78.subsets.rs
/*
 * @lc app=leetcode id=78 lang=rust
 *
 * [78] Subsets
 */

// @lc code=start
impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut subsets = Vec::new();
        fn backtrack(
            nums: &Vec<i32>,
            start: usize,
            subset: &mut Vec<i32>,
            subsets: &mut Vec<Vec<i32>>,
        ) {
            subsets.push(subset.clone());
            for i in start..nums.len() {
                subset.push(nums[i]);
                backtrack(nums, i + 1, subset, subsets);
                subset.pop();
            }
        }
        backtrack(&nums, 0, &mut vec![], &mut subsets);
        subsets
    }
}
// @lc code=end

File: javascript/78.subsets.js
/*
 * @lc app=leetcode id=78 lang=javascript
 *
 * [78] Subsets
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
  const sets = [];
  function backtrack(i, set) {
    sets.push([...set]);
    for (; i < nums.length; i++) {
      set.push(nums[i]);
      backtrack(i + 1, set);
      set.pop();
    }
  }
  backtrack(0, []);
  return sets;
};
// @lc code=end
